### PR TITLE
JA: Translated site/ja/tutorials/customization/performance.ipynb

### PR DESCRIPTION
**Pull request by masa-ita**

### CODE REVIEW

File: site/ja/tutorials/customization/performance.ipynb [hunk 1]
```diff
@@ -0,0 +1,1287 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "J122XQYG7W6w"
+   },
+   "source": [
+    "TensorFlow 2.0 では 先行実行（Eager Execution） が既定で有効になっています。ユーザーインターフェイスは直感的で柔軟です。（一回限りの演算は大幅に簡単かつ迅速です）しかしながら、それは性能と展開の面での犠牲の上に成り立っています。\n",
+    "\n",
+    "最高の性能を上げ、モデルをどこへでも展開できるようにするには、プログラムから計算グラフを作成するために `tf.function` を使います。\n",
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [weak suggestion] `最高の性能を上げ、` -> `最高性能を得ながら`
[weak suggestion] `作成するために` -> `作成する`

File: site/ja/tutorials/customization/performance.ipynb [hunk 2]
```diff
@@ -0,0 +1,1287 @@
...
+    "AutoGraph のおかげで、驚くほど多くの Python コードが tf.function でそのまま動作しますが、気をつけなければならない落とし穴も存在します。\n",
+    "\n",
+    "ポイントと推奨事項は下記の通りです。\n",
+    "\n",
+    "- オブジェクトの変更やリストへの追加などのような Python の副作用に依存しないこと\n",
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [suggestion] `などのような` -> `ような`

File: site/ja/tutorials/customization/performance.ipynb [hunk 3]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "uZ4Do2AV80cO"
+   },
+   "source": [
+    "## トレーシングとポリモーフィズム\n",
+    "\n",
+    "Python の動的型付けは、関数をさまざまな型の引数で呼び出すことができ、Python がそれぞれのシナリオで異なる動作をするということを意味します。\n",
+    "\n",
+    "他方で、TensorFlow の計算グラフでは、dtype と shape の次元が静的であることが必要です。`tf.function` は、適切な計算グラフを生成するために必要なときには関数をトレースし直して、このギャップをつなぐ役割を果たします。\n",
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [suggestion] `適切な` -> `正しい`

Comment (kiszk) [reply site/ja/tutorials/customization/performance.ipynb]: This file may not have a translation of `Most of the subtlety of `tf.function` usage stems from this retracing behavior.`

File: site/ja/tutorials/customization/performance.ipynb [hunk 4]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "4pJqkDR_Q2wz"
+   },
+   "source": [
+    "トレースの動作を制御するためには、下記のようなテクニックを使います。\n",
+    "\n",
+    "- 新しい `tf.function` を作成する。別々の `tf.function` オブジェクトがトレースを共有することはない。\n",
+    "- 特定のトレースを得るには `get_concrete_function` メソッドを使用する。\n",
+    "- 計算グラフの呼び出し時に1回だけトレースをさせるには、 `input_signature` を指定して `tf.function` を呼び出す。"
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [weak suggestion] `トレースをさせる` -> `トレースを行う`

File: site/ja/tutorials/customization/performance.ipynb (hunk 3 above)
Comment (kiszk) [reply site/ja/tutorials/customization/performance.ipynb]: [suggestion] `関数をトレースし直して` -> `再トレースして`  for consistency within a file

File: site/ja/tutorials/customization/performance.ipynb [hunk 5]
```diff
@@ -0,0 +1,1287 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "Es0WZkLIUSdu"
+   },
+   "source": [
+    "## いつサイトレースするのか？\n",
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: `サイトレース` -> `再トレース`

File: site/ja/tutorials/customization/performance.ipynb [hunk 6]
```diff
@@ -0,0 +1,1287 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "AY5oiQN0XIyA"
+   },
+   "source": [
+    "## 引数は Python か？ Tensor か？\n",
+    "\n",
+    "しばしば、ハイパーパラメータやグラフ構成を制御するために Python の引数が使われます。例えば、`num_layers=10` や `training=True` あるいは `nonlinearity='relu'` のようにです。このため、この Python 引数が変更されると、計算グラフを再びトレースする必要があるということになります。\n",
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [weak suggestion] `Python の引数` -> `Python 引数` for consistency within this paragraph.

File: site/ja/tutorials/customization/performance.ipynb [hunk 7]
```diff
@@ -0,0 +1,1287 @@
...
+    "\n",
+    "しかし、グラフの生成を制御するために Python 引数を使用する必要はありません。これらのケースでは、Python引数の値の変更が不必要な再トレースを引き起こす可能性があります。例えば、この訓練ループでは、AutoGraph は動的に巻き戻されます。何回もトレースをしているにも関わらず、生成された計算グラフは全く同一です。これは少し非効率です。"
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: `動的に巻き戻されます` -> `動的に展開を行います`

File: site/ja/tutorials/customization/performance.ipynb [hunk 8]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "129-iRsPS-gY"
+   },
+   "source": [
+    "## `tf.function` の中の副作用\n",
+    "\n",
+    "一般的には、（印字やオブジェクト変更のような）Python の副作用は、トレーシングの最中にだけ発生します。それでは、どうしたら `tf.function` で安定的に副作用を起こすことができるでしょうか？\n",
+    "\n",
+    "一般的な原則は、トレースをデバッグする際には  Python の副作用だけを使用するというものです。そうでなければ、`tf.Variable.assign`、`tf.print`、そして `tf.summary` のような TensorFlow の演算を使うことが、呼び出しの都度 TensorFlowランタイムによってトレースされ実行されることを確実にする方法です。一般には、関数形式を使用することで最も良い結果を得られます。 "
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [strong suggestion] `トレースをデバッグする際には  Python の副作用だけを使用するというものです` -> `トレースをデバッグする際にだけ  Python の副作用を使用するというものです`

File: site/ja/tutorials/customization/performance.ipynb [hunk 9]
```diff
@@ -0,0 +1,1287 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "e1I0dPiqTV8H"
+   },
+   "source": [
+    "`tf.function` が呼び出されるたびに Python のコードを実行したいのであれば、`tf.py_function` がぴったりです。`tf.py_function` の欠点は、ポータブルでないこと、それほど性能が高くないこと、（マルチGPU、TPUの）分散環境ではうまく動作しないことなどです。また、`tf.py_function` 計算グラフに組み込まれるため、入出力すべてをテンソルにキャストします。"
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: `計算グラフに` -> `は計算グラフに`

File: site/ja/tutorials/customization/performance.ipynb [hunk 10]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "msTmv-oyUNaf"
+   },
+   "source": [
+    "## Python の状態に注意\n",
+    "\n",
+    "ジェネレーターやイテレーターなど Python の機能の多くは、状態を追跡するために Python のランタイムに依存しています。これらの仕組みは、一般的には 先行実行（Eager） モードでも期待通りに動作しますが、トレーシング動作のため、 `tf.function` の中では予期しないことが起きることがあります。\n",
+    "\n",
+    "1例として、イテレーターの状態が進行するのは Python の副作用であり、トレーシングの中だけで発生します。"
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [weak suggetion] `進行するのは` -> `進めるのは`

File: site/ja/tutorials/customization/performance.ipynb [hunk 11]
```diff
@@ -0,0 +1,1287 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "5XMGXMu-Ufjm"
+   },
+   "source": [
+    "イテレーターが tf.function の中で生成されすべて使われる場合には、正しく動作するはずです。しかし、イテレーター全体がトレースされることとなり、巨大な計算グラフの生成をまねく可能性があります。これは、望みどおりの動作かもしれません。しかし、もし Python のリストとして表されたメモリー上の巨大なデータセットを使って訓練を行うとすると、これは非常に大きな計算グラフを生成することになり、`tf.function` がスピードアップにはつながらないと考えられます。\n",
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [weak suggestion] `生成されすべて使われる` -> `すべて生成され使われる`

File: site/ja/tutorials/customization/performance.ipynb [hunk 12]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "lPr_6mK_AQWL"
+   },
+   "source": [
+    "## 変数\n",
+    "\n",
+    "`tf.function` の中では、意図したコードの実行順序を活用するという同じアイデアを使って、変数の作成と活用を簡単に行うことができます。しかし、ひとつだけ非常に重要な欠点があります。それは、変数を使った場合、先行実行モードとグラフモードでは動作が変わるコードを書いてしまう可能性があるということです。\n",
+    "\n",
+    "特に、呼び出しの都度新しい変数を作成する場合にこれが発生します。トレーシングの意味では、`tf.function` は呼び出しのたびに同じ変数を再利用しますが、先行実行モードでは呼び出しごとに新しい変数を生成します。この間違いを防止するため、`tf.function` は危険な変数の生成動作を見つけるとエラーを発生させます。"
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [weak suggestion] `意味` -> `意味論` or `セマンティクス`

File: site/ja/tutorials/customization/performance.ipynb [hunk 13]
```diff
@@ -0,0 +1,1287 @@
...
+   "execution_count": null,
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "_IOVc1eujMH2"
+   },
+   "outputs": [],
+   "source": [
+    "# 変数の初期化は、関数の引数や他の変数の値に依存可能\n",
+    "# 制御の依存関係を生成するのと同じ手法で、正しい初期化の順序を推測可能\n",
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [suggestion] `推測` -> `解決` or `発見`

File: site/ja/tutorials/customization/performance.ipynb [hunk 14]
```diff
@@ -0,0 +1,1287 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "@tf.function\n",
+    "def f():\n",
+    "  if tf.constant(True):\n",
+    "    x = tf.ones([3, 3])\n",
+    "  return x\n",
+    "\n",
+    "# Throws an error because both branches need to define `x`.\n",
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: Do we need to translate this line?

File: site/ja/tutorials/customization/performance.ipynb [hunk 15]
```diff
@@ -0,0 +1,1287 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "9gmLpHY-bkly"
+   },
+   "source": [
+    "`tf.cond` と同様に、`tf.while_loop` にも、色々と注意すべき細かな点があります。\n",
+    "\n",
+    "- 繰り返しの実行回数が 0 である可能性があるため、while_loop の後で使用されるテンソルは、繰り返しの前に初期化されなければならない\n",
+    "- 繰り返しの変数はすべて、各繰り返しを通じてその形状と dtype が変わらないことが必要"
```
Comment (kiszk) [inline site/ja/tutorials/customization/performance.ipynb]: [weak suggestion] `繰り返しの変数はすべて` -> `すべての繰り返しの変数は`

File: site/ja/tutorials/customization/performance.ipynb (hunk 1 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thank you for your suggestion.
For the latter I think "`tf.function` を使ってプログラムから計算グラフを作成します。" is better and natural.

File: site/ja/tutorials/customization/performance.ipynb (hunk 2 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.  I fixed it.

File: site/ja/tutorials/customization/performance.ipynb (hunk 3 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thank you. I fixed them.

File: site/ja/tutorials/customization/performance.ipynb (hunk 4 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks．I fixed it.

File: site/ja/tutorials/customization/performance.ipynb (hunk 5 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Oops! Fixed.

File: site/ja/tutorials/customization/performance.ipynb (hunk 6 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.  I fixed it.

File: site/ja/tutorials/customization/performance.ipynb (hunk 7 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thank you.  I fixed.

File: site/ja/tutorials/customization/performance.ipynb (hunk 8 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.  I missed it.

File: site/ja/tutorials/customization/performance.ipynb (hunk 9 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.  Fixed.

File: site/ja/tutorials/customization/performance.ipynb (hunk 10 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.  I changed it to "進むのは".

File: site/ja/tutorials/customization/performance.ipynb (hunk 11 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thank you.  I'd like to leave it.

File: site/ja/tutorials/customization/performance.ipynb (hunk 12 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.  I think "セマンティクス" is better.

File: site/ja/tutorials/customization/performance.ipynb (hunk 13 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thank you.  I'd like to use "発見".

File: site/ja/tutorials/customization/performance.ipynb (hunk 14 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.  Done.

File: site/ja/tutorials/customization/performance.ipynb (hunk 15 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thank you. Fixed.

File: site/ja/tutorials/customization/performance.ipynb [hunk 16]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "tRdlnCfV_UTn"
+   },
+   "source": [
+    "## 制御の依存関係の自動化\n",
+    "\n",
+    "プログラミングモデルとしての関数が一般的なデータフローグラフに対して非常に優位である点は、意図したコードの振る舞いがどのようなものであるかということについて、より多くの情報をランタイムに与えられるということにあります。\n",
+    "\n",
+    "例えば、同じ変数を何度も読んだり書いたりするコードを書く場合、データフローグラフではもともと意図されていた演算の順番を自然に組み込むわけではありません。`tf.function` の中では、もともとの Python コードの文の実行順序を参照することで、実行順序の曖昧さを解消します。これにより、`tf.function` の中のステートフルな演算の順序が、先行実行モードのセマンティクスを模していることになります。\n",
```
Comment (AseiSugiyama) [inline site/ja/tutorials/customization/performance.ipynb]: [weak suggestion]

```
先行実行モード
=====>
Eager モード
```

File: site/ja/tutorials/customization/performance.ipynb [hunk 17]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "lPr_6mK_AQWL"
+   },
+   "source": [
+    "## 変数\n",
+    "\n",
+    "`tf.function` の中では、意図したコードの実行順序を活用するという同じアイデアを使って、変数の作成と活用を簡単に行うことができます。しかし、ひとつだけ非常に重要な欠点があります。それは、変数を使った場合、先行実行モードとグラフモードでは動作が変わるコードを書いてしまう可能性があるということです。\n",
+    "\n",
+    "特に、呼び出しの都度新しい変数を作成する場合にこれが発生します。トレーシングの意味では、`tf.function` は呼び出しのたびに同じ変数を再利用しますが、先行実行モードでは呼び出しごとに新しい変数を生成します。この間違いを防止するため、`tf.function` は危険な変数の生成動作を見つけるとエラーを発生させます。"
```
Comment (AseiSugiyama) [inline site/ja/tutorials/customization/performance.ipynb]: [weak suggestion]

```
先行実行モード
=====>
Eager モード
```

File: site/ja/tutorials/customization/performance.ipynb [hunk 18]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "5f05Vr_YBUCz"
+   },
+   "source": [
+    "# AutoGraph の使用\n",
+    "\n",
+    "[autograph](https://www.tensorflow.org/guide/autograph) ライブラリは `tf.function` に完全に統合されており、計算グラフの中で動的に実行される条件文や繰り返しを書くことができます。\n",
+    "\n",
+    "`tf.cond` や `tf.while_loop` は `tf.function` でも使えるが、制御の流れを含むコードは、命令形式で書いたほうが書きやすいし理解しやすいです。"
```
Comment (AseiSugiyama) [inline site/ja/tutorials/customization/performance.ipynb]: [typo]

```
使えるが、
=====>
使えますが、
```

File: site/ja/tutorials/customization/performance.ipynb [hunk 19]
```diff
@@ -0,0 +1,1287 @@
...
+    "TensorFlow 2.0 では 先行実行（Eager Execution） が既定で有効になっています。ユーザーインターフェイスは直感的で柔軟です。（一回限りの演算は大幅に簡単かつ迅速です）しかしながら、それは性能と展開の面での犠牲の上に成り立っています。\n",
+    "\n",
+    "最高性能を得ながら、モデルをどこへでも展開できるようにするには、`tf.function` を使ってプログラムから計算グラフを作成します。\n",
+    "AutoGraph のおかげで、驚くほど多くの Python コードが tf.function でそのまま動作しますが、気をつけなければならない落とし穴も存在します。\n",
+    "\n",
+    "ポイントと推奨事項は下記の通りです。\n",
+    "\n",
+    "- オブジェクトの変更やリストへの追加のような Python の副作用に依存しないこと\n",
+    "- tf.function は、NumPy の演算や Python のプリミティブよりも TensorFlow の演算に適している\n",
+    "- 迷ったときは、`for x in y` という常套句を使うこと"
```
Comment (AseiSugiyama) [inline site/ja/tutorials/customization/performance.ipynb]: I guess `イディオム` is better than `常套句` as you translated below.

File: site/ja/tutorials/customization/performance.ipynb [hunk 20]
```diff
@@ -0,0 +1,1287 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "J122XQYG7W6w"
+   },
+   "source": [
+    "TensorFlow 2.0 では 先行実行（Eager Execution） が既定で有効になっています。ユーザーインターフェイスは直感的で柔軟です。（一回限りの演算は大幅に簡単かつ迅速です）しかしながら、それは性能と展開の面での犠牲の上に成り立っています。\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thank you for your review.
The official TensorFlow web site uses "積極的実行" as a translated word for "Eager Execution".
But there was a discussion that we Japanese engineers use words "先行/遅延" as translated words for "eager/delated", so I tried "先行実行”.
As there is no consensus now, I revert it to "Eager Execution".

File: site/ja/tutorials/customization/performance.ipynb [hunk 21]
```diff
@@ -0,0 +1,1287 @@
...
+    "\n",
+    "最高性能を得ながら、モデルをどこへでも展開できるようにするには、`tf.function` を使ってプログラムから計算グラフを作成します。\n",
+    "AutoGraph のおかげで、驚くほど多くの Python コードが tf.function でそのまま動作しますが、気をつけなければならない落とし穴も存在します。\n",
+    "\n",
+    "ポイントと推奨事項は下記の通りです。\n",
+    "\n",
+    "- オブジェクトの変更やリストへの追加のような Python の副作用に依存しないこと\n",
+    "- tf.function は、NumPy の演算や Python のプリミティブよりも TensorFlow の演算に適している\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.  I agree.

File: site/ja/tutorials/customization/performance.ipynb [hunk 22]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "D25apou9IOXa"
+   },
+   "outputs": [],
+   "source": [
+    "import contextlib\n",
+    "\n",
+    "# 遭遇するかもしれないいくつかのエラーをデモするためのヘルパープログラム\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks．I changed it.

File: site/ja/tutorials/customization/performance.ipynb [hunk 23]
```diff
@@ -0,0 +1,1287 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "rfayNj-ZIkIB"
+   },
+   "source": [
+    "あなたが定義した `tf.function` は TensorFlow の中核となる演算のようなものです。即時実行することも、グラフの中で使うこともでき、勾配も持っているなどです。"
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.  These are much better.

File: site/ja/tutorials/customization/performance.ipynb [hunk 24]
```diff
@@ -0,0 +1,1287 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "SbtT1-Wm70F2"
+   },
+   "outputs": [],
+   "source": [
+    "# 演算のような関数の例\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: I got it thank you!

File: site/ja/tutorials/customization/performance.ipynb [hunk 25]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "129-iRsPS-gY"
+   },
+   "source": [
+    "## `tf.function` の中の副作用\n",
+    "\n",
+    "一般的には、（印字やオブジェクト変更のような）Python の副作用は、トレーシングの最中にだけ発生します。それでは、どうしたら `tf.function` で安定的に副作用を起こすことができるでしょうか？\n",
+    "\n",
+    "一般的な原則は、トレースをデバッグする際にだけ Python の副作用を使用するというものです。そうでなければ、`tf.Variable.assign`、`tf.print`、そして `tf.summary` のような TensorFlow の演算を使うことが、呼び出しの都度 TensorFlowランタイムによってトレースされ実行されることを確実にする方法です。一般には、関数形式を使用することで最も良い結果を得られます。 "
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Yes!

File: site/ja/tutorials/customization/performance.ipynb [hunk 26]
```diff
@@ -0,0 +1,1287 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "msTmv-oyUNaf"
+   },
+   "source": [
+    "## Python の状態に注意\n",
+    "\n",
+    "ジェネレーターやイテレーターなど Python の機能の多くは、状態を追跡するために Python のランタイムに依存しています。これらの仕組みは、一般的には 先行実行（Eager） モードでも期待通りに動作しますが、トレーシング動作のため、 `tf.function` の中では予期しないことが起きることがあります。\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks. I fixed.

File: site/ja/tutorials/customization/performance.ipynb (hunk 16 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: It's OK.

File: site/ja/tutorials/customization/performance.ipynb (hunk 17 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.

File: site/ja/tutorials/customization/performance.ipynb (hunk 18 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks. I fixed it.

File: site/ja/tutorials/customization/performance.ipynb [hunk 27]
```diff
@@ -0,0 +1,1287 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "tRdlnCfV_UTn"
+   },
+   "source": [
+    "## 制御の依存関係の自動化\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: I agree!

File: site/ja/tutorials/customization/performance.ipynb [hunk 28]
```diff
@@ -0,0 +1,1287 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "SASm0ss8erVX"
+   },
+   "outputs": [],
+   "source": [
+    "# 制御の依存関係の自動化\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.

File: site/ja/tutorials/customization/performance.ipynb (hunk 18 above)
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.

File: site/ja/tutorials/customization/performance.ipynb [hunk 29]
```diff
@@ -0,0 +1,1287 @@
...
+   "metadata": {
+    "colab_type": "text",
+    "id": "yho4J0a0ZkQS"
+   },
+   "source": [
+    "## AutoGraph と繰り返し\n",
+    "\n",
+    "AutoGraph には繰り返しの変換にいくつかの単純なルールがあります。\n",
+    "\n",
+    "- `for`: イテラブルなものがテンソルである場合に変換する\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.

File: site/ja/tutorials/customization/performance.ipynb [hunk 30]
```diff
@@ -0,0 +1,1287 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "dSr64Xn6ap-S"
+   },
+   "source": [
+    " 繰り返しに、テンソルに依存する `break` や、途中での `return` がある場合、一番外側の条件あるいはイテラブルはテンソルである必要があります。 \n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.

File: site/ja/tutorials/customization/performance.ipynb [hunk 31]
```diff
@@ -0,0 +1,1287 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@tf.function\n",
+    "def buggy_loop_type_changes():\n",
+    "  x = tf.constant(0, dtype=tf.float32)\n",
+    "  for i in tf.range(3): # Yields tensors of type tf.int32...\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thank you for suggestion.

File: site/ja/tutorials/customization/performance.ipynb (hunk 19 above)
Comment (masa-ita) [reply site/ja/tutorials/customization/performance.ipynb]: Thanks.

**Comment by AseiSugiyama** [comment]
@masa-ita Thank you for your work, LGTM!

**Pull request merged**